pkg/operator/controller/controller_lb.go

```diff
@@ -0,0 +1,161 @@
+package controller
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/sirupsen/logrus"
+
+	ingressv1alpha1 "github.com/openshift/cluster-ingress-operator/pkg/apis/ingress/v1alpha1"
+	"github.com/openshift/cluster-ingress-operator/pkg/manifests"
+	"github.com/openshift/cluster-ingress-operator/pkg/util/slice"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+
+	configv1 "github.com/openshift/api/config/v1"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+)
+
+const (
+	// loadBalancerServiceFinalizer is applied to load balancer services to ensure
+	// we can manage deletion of associated DNS records.
+	loadBalancerServiceFinalizer = "ingress.openshift.io/operator"
+
+	// awsLBProxyProtocolAnnotation is used to enable the PROXY protocol on any
+	// AWS load balancer services created.
+	awsLBProxyProtocolAnnotation = "service.beta.kubernetes.io/aws-load-balancer-proxy-protocol"
+)
+
+// ensureLoadBalancerService creates an LB service if one is desired but absent.
+// Always returns the current LB service if one exists (whether it already
+// existed or was created during the course of the function).
```

Comment:
This means `ensureRouterForIngress` will `ensureDNS` an existing LB service even if none is desired, right? Do we want to do that?

Reply:
If one exists at all, then `ensureDNS` will try and create any DNS for it. If later on the clusteringress is deleted, we'll tear it all down.

If the clusteringress was deleted we won't even be in this code path in the first place.

Not sure there's an error here yet. However, I think this decomposition lets us at least start reasoning about the scenarios...